### DISCUSSION TITLE
Assignment Variable Selection Based on Conditional Check Leads to Step Failure

### DISCUSSION THREAD

**Discussion by kumarjitu21** [community/community#68217] [Actions] [answered]
### Select Topic Area

Question

### Body

Can someone please help me to point out the issue that i encounter while executing the conditional check using the below code:

```
jobs:
  job1:
    steps:
      - name: get tag value
        run: |
          if [[ "${{ github.ref }}" == "refs/heads/develop" ]]; then
            deploy_env="dev"
          elif [[ "${{ github.ref }}" == "refs/heads/main" ]]; then
            deploy_env="prod"
          elif [[ ${{ contains(github.ref, 'refs/heads/release/') }} || ${{ contains(github.ref, 'refs/heads/hotfix/') }} ]]; then
            deploy_env="qa"
          else
            deploy_env="latest"
          fi
          echo "deploy_env :: ${deploy_env}"
```
          
          
In execution logs, it always printing "deploy_env :: latest" even if github.ref = "refs/heads/develop".

**Comment by GerychCode** [comment]
The issue with your code is that you are using the contains() function incorrectly. The contains() function takes two arguments: a string and a substring. It returns true if the substring is found in the string, and false otherwise.

In your code, you are using the contains() function to check if the GitHub ref starts with the string "refs/heads/release/" or "refs/heads/hotfix/". However, you are using the || operator to combine the two conditions, which means that either condition will evaluate to true if the GitHub ref starts with either string.

To fix the issue, you should use the and operator to combine the two conditions. This will ensure that both conditions must be met for the deploy_env variable to be set to "qa".

```
jobs:
job1:
steps:
- name: get tag value
run: |
  if [[ "${{ github.ref }}" == "refs/heads/develop" ]]; then
    deploy_env="dev"
  elif [[ "${{ github.ref }}" == "refs/heads/main" ]]; then
    deploy_env="prod"
  elif [[ ${{ contains(github.ref, 'refs/heads/release/') }} && ${{ contains(github.ref, 'refs/heads/hotfix/') }} ]]; then
    deploy_env="qa"
  else
    deploy_env="latest"
  fi
  echo "deploy_env :: ${deploy_env}"_
```
With this corrected code, the deploy_env variable will be set to "dev" if the GitHub ref is "refs/heads/develop".

**Reply by kumarjitu21** [reply]
Thanks @GerychCode for the response. Actually i would like to check either condition for deploy_env="qa", if i use && operator it will never match for qa tag.

**Comment by seyLu** [comment]
#### Hi @kumarjitu21,

I've test run your workflow using:
```yaml
name: Test

on:
  push:

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - name: test run
        run: |
          if [[ "${{ github.ref }}" == "refs/heads/develop" ]]; then
            deploy_env="dev"
          elif [[ "${{ github.ref }}" == "refs/heads/main" ]]; then
            deploy_env="prod"
          elif [[ ${{ contains(github.ref, 'refs/heads/release/') }} || ${{ contains(github.ref, 'refs/heads/hotfix/') }} ]]; then
            deploy_env="qa"
          else
            deploy_env="latest"
          fi
          echo "deploy_env :: ${deploy_env}"
```

- on `main` branch:
![image](https://github.com/community/community/assets/98249191/3d13e99b-596f-40fe-9546-e827a524b73e)

- on `develop` branch:
![image](https://github.com/community/community/assets/98249191/c7dceecd-918c-4458-9e72-0683da033a96)

- on `release/*` or `hotfix/*` branch:
![image](https://github.com/community/community/assets/98249191/01ba78f4-c33f-4194-a337-68d6dbdc0e7b)
![image](https://github.com/community/community/assets/98249191/6d932516-bc64-4492-87f0-f6905f86678f)

**Reply by kumarjitu21** [reply]
@seyLu The anticipated sequence of actions is as you described, but the puzzling aspect is that it does not function identically in my CI runner. Instead, it consistently assigns deploy_env as "latest." Additionally, it generates the following warnings:

```
_/__w/_temp/cb6a354a-33bc-4be3-8861-64313ef5ab2d.sh: 3: [[: not found
/__w/_temp/cb6a354a-33bc-4be3-8861-64313ef5ab2d.sh: 5: [[: not found
/__w/_temp/cb6a354a-33bc-4be3-8861-64313ef5ab2d.sh: 7: [[: not found
```

It appeared that runner has default sh command-line interpreter instead of bash, So I tried to rewrite the condition check in different ways like below:

```
if [ ${{ github.ref }} = "refs/heads/develop" ]; then
	deploy_env="dev"
elif [ ${{ github.ref }} = "refs/heads/main" ]; then
	deploy_env="prod"
elif [[ ${{ contains(github.ref, 'refs/heads/release/') }} || ${{ contains(github.ref, 'refs/heads/hotfix/') }} ]]; then
	deploy_env="qa"
else
	deploy_env="latest"
fi
echo "deploy_env :: ${deploy_env}"
```

The code mentioned above successfully handled the first two conditions, but it encountered an issue with the third check, leading to the assignment of deploy_env=latest when that condition is met. I would appreciate if you have any suggestions/inputs on this.

**Reply by seyLu** [reply]
You did not specify the runner for the job (can be ubuntu, mac, windows, or self-hosted).
![image](https://github.com/community/community/assets/98249191/91909eb1-c428-4862-8e3b-0111cd7e03ac)


You can read more on: [Github docs: Choosing a runner for a job](https://docs.github.com/en/actions/using-jobs/choosing-the-runner-for-a-job)

**Reply by kumarjitu21** [reply]
I've added only actual code snippet here and Obviously, I'm using runner here using `runs-on: self-hosted`.

**Reply by seyLu** [reply, marked as answer]
> Obviously, I'm using runner here using runs-on: self-hosted.

Ummm, how would I know? Please be more explicit next time and provide more context on snippets -- devs are not mind readers.

Anyways, you can specify the shell to use on the step:

```yaml
jobs:
  test:
    runs-on: self-hosted
    steps:
      - name: test run
        shell: bash
        run: ...
```

You can read more on: [Github docs: `[jobs.<job_id>.steps[*].shell]`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell)

**Reply by kumarjitu21** [reply]
`shell: bash` did the trick. I understand that developers can't read minds, but regardless, I appreciate your assistance and thank you for sharing the link to the documentation. I must have missed that when I was going through it.
